apollo-router/src/layers/async_checkpoint.rs

```diff
@@ -67,6 +68,83 @@ where
     }
 }
 
+/// [`Service`] for OneShot (single use) Asynchronous Checkpoints. See [`ServiceBuilderExt::oneshot_checkpoint_async()`](crate::layers::ServiceBuilderExt::oneshot_checkpoint_async()).
+#[allow(clippy::type_complexity)]
+pub struct OneShotAsyncCheckpointService<S, Fut, Request>
+where
+    Request: Send + 'static,
+    S: Service<Request, Error = BoxError> + Send + 'static,
+    <S as Service<Request>>::Response: Send + 'static,
+    <S as Service<Request>>::Future: Send + 'static,
+    Fut: Future<Output = Result<ControlFlow<<S as Service<Request>>::Response, Request>, BoxError>>,
+{
+    inner: Option<S>,
+    checkpoint_fn: Arc<Pin<Box<dyn Fn(Request) -> Fut + Send + Sync + 'static>>>,
+}
+
+impl<S, Fut, Request> OneShotAsyncCheckpointService<S, Fut, Request>
+where
+    Request: Send + 'static,
+    S: Service<Request, Error = BoxError> + Send + 'static,
+    <S as Service<Request>>::Response: Send + 'static,
+    <S as Service<Request>>::Future: Send + 'static,
+    Fut: Future<Output = Result<ControlFlow<<S as Service<Request>>::Response, Request>, BoxError>>,
+{
+    /// Create an `OneShotAsyncCheckpointLayer` from a function that takes a Service Request and returns a `ControlFlow`
+    pub fn new<F>(checkpoint_fn: F, service: S) -> Self
+    where
+        F: Fn(Request) -> Fut + Send + Sync + 'static,
+    {
+        Self {
+            checkpoint_fn: Arc::new(Box::pin(checkpoint_fn)),
+            inner: Some(service),
+        }
+    }
+}
+
+impl<S, Fut, Request> Service<Request> for OneShotAsyncCheckpointService<S, Fut, Request>
+where
+    Request: Send + 'static,
+    S: Service<Request, Error = BoxError> + Send + 'static,
+    <S as Service<Request>>::Response: Send + 'static,
+    <S as Service<Request>>::Future: Send + 'static,
+    Fut: Future<Output = Result<ControlFlow<<S as Service<Request>>::Response, Request>, BoxError>>
+        + Send
+        + 'static,
+{
+    type Response = <S as Service<Request>>::Response;
+
+    type Error = BoxError;
+
+    type Future = BoxFuture<'static, Result<Self::Response, Self::Error>>;
+
+    fn poll_ready(
+        &mut self,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Result<(), Self::Error>> {
+        // Return an error if we no longer have an inner service
+        match self.inner.as_mut() {
+            Some(inner) => inner.poll_ready(cx),
+            None => Poll::Ready(Err("One shot must only be called once".into())),
+        }
+    }
+
+    fn call(&mut self, req: Request) -> Self::Future {
+        let checkpoint_fn = Arc::clone(&self.checkpoint_fn);
+        let inner = self
+            .inner
+            .take()
+            .expect("One shot must only be called once");
```

Comment:
same as with poll_ready, could we return a boxerror instead ?

Reply:
https://github.com/apollographql/router/pull/3878

Reply:
We could. This merged yesterday, but I could do a follow-up to change:
```
     fn call(&mut self, req: Request) -> Self::Future {
         let checkpoint_fn = Arc::clone(&self.checkpoint_fn);
-        let inner = self
-            .inner
-            .take()
-            .expect("One shot must only be called once");
+        let inner = match self.inner.take() {
+            Some(inner) => inner,
+            None => {
+                return Box::pin(async move { Err("No inner service found".into()) });
+            }
+        };
         Box::pin(async move {
             match (checkpoint_fn)(req).await {
                 Ok(ControlFlow::Break(response)) => Ok(response),
```